File: Common/Action/src/main/java/com/greentree/action/observable/ObjIntObservable.java
package com.greentree.action.observable;

import java.util.function.Consumer;
import java.util.function.IntConsumer;
import java.util.function.ObjIntConsumer;

import com.greentree.action.ListenerCloser;

public interface ObjIntObservable<T> extends RunObservable, IntObservable, ObjectObservable<T> {
	
	ObjIntObservable<?> NULL = new ObjIntObservable<>() {
		
		private static final long serialVersionUID = 1L;
		
		@Override
		public ListenerCloser addListener(ObjIntConsumer<? super Object> listener) {
			return ListenerCloser.EMPTY;
		}
		
		@Override
		public int listenerSize() {
			return 0;
		}
		
	};
	
	@SuppressWarnings("unchecked")
	static <T> ObjIntObservable<T> getNull() {
		return (ObjIntObservable<T>) NULL;
	}
	
	@Override
	default ListenerCloser addListener(Runnable listener) {
		return addListener((e, i)->listener.run());
	}
	
	@Override
	default ListenerCloser addListener(Consumer<? super T> listener) {
		return addListener((e, i)->listener.accept(e));
	}
	
	@Override
	default ListenerCloser addListener(IntConsumer listener) {
		return addListener((e, i)->listener.accept(i));
	}
	
	ListenerCloser addListener(ObjIntConsumer<? super T> listener);
	
}

File: Commons/Action/src/main/java/com/greentree/commons/action/observable/ObjIntObservable.kt
package com.greentree.commons.action.observable

import com.greentree.commons.action.ListenerCloser
import java.util.function.IntConsumer
import java.util.function.ObjIntConsumer

interface ObjIntObservable<T> : RunObservable, IntObservable, ObjectObservable<T> {

	override fun addListener(listener: Runnable): ListenerCloser {
		return addListener { _, _ -> listener.run() }
	}

	fun addListener(listener: ObjIntConsumer<in T>): ListenerCloser

	override fun addListener(listener: (T) -> Unit): ListenerCloser {
		return addListener { e: T, _ -> listener.invoke(e) }
	}

	override fun addListener(listener: IntConsumer): ListenerCloser {
		return addListener { _, i: Int -> listener.accept(i) }
	}

	data object Empty : ObjIntObservable<Any?> {

		fun <T> instance() = Empty as ObjIntObservable<T>

		override val listenerCount: Int
			get() = 0

		override fun addListener(listener: ObjIntConsumer<in Any?>): ListenerCloser {
			return ListenerCloser.Empty
		}
	}
}
